File: C++/2021/day3/solution.cpp
#include "utils.h"

using namespace std;

using report_t = vector<vector<int>>;

report_t read_report(const vector<string> &lines) {
  report_t result;

  for (auto &line : lines) {
    vector<int> row;
    for (auto &c : line) {
      row.push_back(c - '0');
    }
    result.push_back(row);
  }

  return result;
}

int to_decimal(vector<int> byte) {
  int result = 0, n = byte.size();
  for (auto i = 0; i < n; i++) {
    result += byte[i] * ipow(2, n - 1 - i);
  }

  return result;
}

int scan_numbers(const report_t &current, int column, int preferred) {
  if (current.size() == 1)
    return to_decimal(current.front());

  if (column >= current[0].size())
    return 0;

  int r = current.size();

  vector<int> ones, zeros;

  for (auto i = 0; i < r; i++) {
    if (current[i][column])
      ones.push_back(i);
    else
      zeros.push_back(i);
  }

  vector<int> selected;
  auto s1 = ones.size(), s0 = zeros.size();

  if (preferred) {
    selected = (s1 == s0 || s1 > s0) ? ones : zeros;
  } else {
    selected = (s1 == s0 || s1 > s0) ? zeros : ones;
  }

  report_t next;
  for (auto &index : selected) {
    next.push_back(current[index]);
  }

  return scan_numbers(next, column + 1, preferred);
}

vector<string> parse_input(const string &file_name) {
  vector<string> ret;

  ifstream fs(file_name);
  string line;
  while (getline(fs, line)) {
    ret.push_back(line);
  }
  return ret;
}

int part1(const report_t &report) {
  int r = report.size(),    //
      c = report[0].size(), //
      h = r / 2;

  vector<int> gamma, epsilon;
  for (auto j = 0; j < c; j++) {
    int sum = 0;
    for (auto i = 0; i < r; i++) {
      sum += report[i][j];
    }
    gamma.push_back(sum > h);
    epsilon.push_back(sum < h);
  }

  return to_decimal(gamma) * to_decimal(epsilon);
}

int part2(const report_t &report) {
  int o2 = scan_numbers(report, 0, 1), //
      co2 = scan_numbers(report, 0, 0);
  return o2 * co2;
}

int main(int argc, char *argv[]) {
  auto lines = parse_input(argv[1]);
  auto report = read_report(lines);

  auto r1 = part1(report);
  println("Part 1: ", r1);

  auto r2 = part2(report);
  println("Part 2: ", r2);

  return 0;
}

File: Go/2021/day3/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {

	if len(os.Args) < 2 {
		fmt.Println("Please provide an input file")
		return
	}

	file, err := os.Open(os.Args[1])

	if err != nil {
		fmt.Println("Couldn't open file input file", err)
		return
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	lines := []string{}
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	fmt.Printf("Part 1: %d\n", part1(lines))
	fmt.Printf("Part 1: %d\n", part2(lines))
}

func part1(lines []string) int {
	var (
		gamma   []byte
		epsilon []byte
	)

	for i := 0; i < len(lines[0]); i++ {
		var (
			ones, zeros int
		)
		for _, line := range lines {
			if line[i] == '1' {
				ones++
			} else {
				zeros++
			}
		}
		if ones > zeros {
			gamma = append(gamma, '1')
			epsilon = append(epsilon, '0')
		} else {
			gamma = append(gamma, '0')
			epsilon = append(epsilon, '1')
		}
	}

	return toDecimal(string(gamma)) * toDecimal(string(epsilon))
}

func part2(lines []string) int {
	oxygen := filter(lines, 0, '1')
	co2 := filter(lines, 0, '0')

	return toDecimal(oxygen) * toDecimal(co2)
}

func filter(lines []string, index int, value byte) string {
	if len(lines) == 1 {
		return lines[0]
	}

	var (
		ones, zeros []string
	)

	for _, line := range lines {
		if line[index] == '1' {
			ones = append(ones, line)
		} else {
			zeros = append(zeros, line)
		}
	}

	if len(ones) >= len(zeros) {
		return filter(ones, index+1, value)
	} else {
		return filter(zeros, index+1, value)
	}
}

func toDecimal(binary string) int {
	decimal := 0
	for i, c := range binary {
		if c == '1' {
			decimal += 1 << (len(binary) - i - 1)
		}
	}
	return decimal
}
